### DISCUSSION TITLE
One-off job pattern

### DISCUSSION THREAD

**Discussion by djeedai** [bevyengine/bevy#3131] [Q&A]
# Motivational example

My game needs to load a configuration `.json` file toward the beginning of the game, then parse it and load other assets depending on what's inside the `.json` configuration. Many other things cannot happen before this is finished (think of it as a loading screen and preparing assets for what's next). I _think_ that this is a reasonably common use case.

Because `AssetServer` is asynchronous in nature (see #3129), it's impossible in general to run the following in sequence:
```rust
let handle = asset_server.load("config.json");      // enqueue async loading job
let json_asset = json_assets.get(&handle).unwrap(); // panic! async loading not done yet in general
// use JSON...
```

_Side note:_ I know I can use directly `File::open()` and workaround the issue by having a blocking load. However this is:
- bad practice: long operations like I/O should be async to prevent UI freeze
- impractical: this complicates a _lot_ running on wasm, where `File::open()` doesn't work
- arguably bad design: two different codepaths for loading assets, one via `AssetServer` and one via direct loading, one of which Bevy knows about and the other it doesn't

Instead, it seems the recommended way according to [the `custom_asset.rs` example](https://github.com/bevyengine/bevy/blob/main/examples/asset/custom_asset.rs) is to have a system that polls for the loading state and eventually when the asset is loaded use it.

# Generalization

The above asset loading pattern is not only quite a lot a boilerplate code to write, it seems that it will leave registered a system that forever will be ticked (or, as an optimization, has a `RunCriteria` that will get checked, but that's marginally better) each frame for the rest of the application lifetime.

More generally, it seems many one-off jobs like this leave behind them a system that is no longer of any use. This can be optimized I believe by having a dedicated `State` for the job or set of jobs, but the state itself is also implemented as a `RunCriteria`, so will still leave behind that check every frame forever.

I suspect that the rationale behind that design is that with a single global state enum (or a few of them), the number of per-frame checks is low. Nonetheless it's 1) non-zero; 2) as far as I know not really well documented as a pattern so may have limitations or drawbacks I don't know; ~~3) possibly moot since there's some talk about removing `Stage` completely (#1375, #2801)~~ [EDIT: `Stage` but not `State`].

# Problem statement

What is a recommended pattern for one-off jobs without impacting too much performance and cluttering the schedule graph? And ideally without having to write too much boilerplate too. Should/could we allow un-registering a system from the ECS scheduler?

**Comment by rparrett** [comment]
My recommendation would be to not worry about the performance impact of the scheduler checking the app's current state. You're going to want to use states in your app anyway, and the cost of this weighed against the rest of the logic in your app will be extremely small.

A bit of confusion regarding `Stage` and `State` is pretty natural, and I don't think that the plans/experiments around a stageless scheduler are super relevant here.

These links would be good starting points.

https://github.com/bevyengine/bevy/blob/main/examples/ecs/state.rs
https://bevy-cheatbook.github.io/programming/states.html

**Reply by djeedai** [reply]
Thanks for the answer. I fixed the original question; I know the difference between `Stage` and `State`, and I'm already using both, not sure why I mixed them. I've already read those links too.

Anyway the point is not only about performance, but also about the clarity of what is running when, and having meaningful states not only for the sake of working around some technical limitation. For example I just added another state value to my main state because I had to wait for some resource to load (because `AssetServer` is async; see #3129); there's no other good reason to have that value, and it makes reasoning more complex. Alternatively I could have added a "sub-state" (a state only used when another state has a particular value, _e.g._ what's the state of the main menu when the app state is `MainMenu`), but I'm not convinced that it's a good approach either.

In short, I'm still not convinced what's a sensible pattern to deal with all those one-off jobs that one might have in a game. ECS is very much oriented to deal with large number of objects by design, so it makes reasoning about this almost as an edge case I find.

**Comment by copygirl** [comment]
Once assets become entities, surely it'll be possible to use an observer to detect when an asset has finished loading and then run your code just once? Or is there already a mechanic to do something similar today?

**Reply by djeedai** [reply]
You need to poll if I remember correctly, there's no event about asset being loaded (is there?). Anyway detecting when the asset is loaded is only partially the issue; it's more of a motivational example for the question. The core issue is that the ECS design Bevy uses with forever registered systems means as your app increases in size, so does the overhead of all registered systems running all the time each frame. The situation has a bit improved since 2021, with one-shot systems in particular, and observers allow hooking into lifecycle without registering a proper system, but it remains that in general a game has lower level systems running on many objects (entities), which ECS is very well suited for, and on the other hand of the spectrum also has plenty of "actions" running very occasionally in response to other things, especially around gameplay, and operating generally on a single entity (most often, the player, or a particular object or NPC), for which ECS is absolutely not a good fit. Other engines like Unity started from the object-centric end of the spectrum, while Bevy started from the opposite, data-driven end, but the reality is that apps need both, and there's no real recommendations / best practice guide on how to handle those "singleton" systems in Bevy. I've found that to be the most difficult part of writing a non-trivial app with Bevy. Extrapolating from what I've done and seen in the past 4 years, I'd imagine a full-featured game would have thousands of systems. But I admit I don't have first-hand experience on an actual large-size codebase using Bevy.
